rfcs/0007-handling-corruption.md

```diff
@@ -0,0 +1,324 @@
+  recover as much of the WAL as possible, rather than truncating the remaining
+  entries.
+
+### Proposed Solution
+I am working under the assumption that, similar to other embeddable key-value
+stores, SlateDB is intended to serve as the low-level durability layer. This
+layer allows developers to build complex logical durability and availability
+systems, such as partitions, replication, and backups.
+
+As mentioned earlier, determining whether the developer or operator prioritizes
+durability or availability is beyond the scope of SlateDB. Therefore, this
+proposal assumes that while maintaining availability is valuable, SlateDB
+should also provide actionable errors and recovery tools for those who
+prioritize durability over availability.
+
+### Handling WAL Corruption
+It is reasonable for a developer to assume that once data is confirmed to be
+written to the WAL, the data or transaction is considered durably written to
+storage. However, as discussed in the "Corruption Review" section of this
+document, this assumption is only theoretical. In practice, it is possible for
+WAL entries written to storage to be corrupted, and this corruption may only be
+discovered during compaction or recovery processes.
+
+It is important to note that during normal LSM operation, the movement of
+entries from the WAL to the SSTable occurs in memory. The WAL is only used as a
+recovery mechanism in case of a catastrophic failure of the LSM. Therefore, the
+loss of committed WAL data can only occur under the following two conditions:
+
+1. The LSM application exits before compaction can complete.
+2. Corruption of the WAL occurs.
+
+Although the probability of both conditions being true is very low, especially
+in the context of durable object storage, it is not impossible. Given this low
+probability, we should avoid complicating the WAL by coupling `put` calls to
+main SSTable writes. The `WriteOptions.AwaitFlush` (Golang) write option should
+be sufficient confirmation that a write or transaction is durably written to
+the WAL.
+
+SlateDB can improve over other implementations during the recovery process. In
+the implementations I tested, WAL recovery occurs asynchronously after the
+database is opened. As a result, the only way to warn the developer of WAL
+corruption is via a logged warning.
+
+To support operators who value durability and to ensure durability by default,
+I propose that `DB::open_with_opts()` perform WAL recovery synchronously. This
+method will return any WAL corruption found as errors and abort the database
+open process until the corruption is resolved. The error returned should be a
+distinct type that can easily be matched to determine if the error is due to
+non-transient corruption or a transient error (such as network connectivity
+issues).
+
+To support developers who value availability, we will introduce a new database
+option called `DbOptions.wal_auto_recovery`, similar in naming to
```

Comment:
i guess this can be handled at the user-side? like if we've got an error on WAL corruption error, we can choose call `RepairDB()` by ourselves to save us back, with understood the risks of data loss on this operation.

leveldb offered a `RepairDB()` operation for this:

https://github.com/google/leveldb/blob/main/db/repair.cc#L446C8-L446C16